vault-tool/src/Network/VaultTool/TOTP.hs

```diff
@@ -0,0 +1,252 @@
+    toJSON = \case
+        SHA1 -> String "SHA1"
+        SHA256 -> String "SHA256"
+        SHA512 -> String "SHA512"
+
+data DigitCount = SixDigits | EightDigits
+    deriving (Show, Eq)
+
+instance FromJSON DigitCount where
+    parseJSON (Number 6) = pure SixDigits
+    parseJSON (Number 8) = pure EightDigits
+    parseJSON x = fail $ "Expected 6 or 8 but received " <> show x
+
+instance ToJSON DigitCount where
+    toJSON = \case
+        SixDigits -> Number 6
+        EightDigits -> Number 8
+
+data Skew = NoSkew | OnePeriodSkew
+    deriving (Show, Eq)
+
+instance ToJSON Skew where
+    toJSON = \case
+        NoSkew -> Number 0
+        OnePeriodSkew -> Number 1
+
+-- | The newly generated Vault TOTP key
+data GeneratedKey = GeneratedKey
+    { -- | The resulting base64 encoded QR code PNG
+      gkrBarcode :: Text,
+      -- | The resulting otpauth url for the key
+      gkrUrl :: Text
+    }
+    deriving (Show, Eq)
+
+instance FromJSON GeneratedKey where
+    parseJSON = withObject "GeneratedKey" $ \v ->
+        GeneratedKey <$> v .: "barcode" <*> v .: "url"
+
+-- | Generates a new TOTP code via Vault's TOTP API
+generateKey :: VaultConnection Authenticated -> VaultMountedPath -> GenerateKeyRequest -> IO GeneratedKey
+generateKey conn path req = fmap unDataWrapper
+    . runVaultRequestAuthenticated conn
+    . newPostRequest (mkPathWithKey KeysNamespace path $ gkrKeyName req)
+    $ Just req
+
+-- | A TOTP key managed in Vault
+data Key = Key
+    {
+      -- | The name of the account associated with the key
+      kAccountName :: Text,
+      -- | The hashing algorithm to use. Options are 'SHA1', 'SHA256', 'SHA512'. (default: 'SHA1')
+      kAlgorithm :: HashAlgorithm,
+      -- | The number of digits in the generated TOTP code. Options are 'SixDigits' or 'EightDigits'. (default: 'SixDigits')
+      kDigitCount :: DigitCount,
+      -- | The name of the key's issuing organization
+      kIssuer :: Text,
+      -- | The length of time in seconds used to generate a counter for the TOTP code calculation (default: 30)
+      kPeriod :: Int
+    }
+    deriving (Show, Eq)
+
+instance FromJSON Key where
+    parseJSON = withObject "Key" $ \v -> Key
+        <$> v .: "account_name"
+        <*> v .: "algorithm"
+        <*> v .: "digits"
+        <*> v .: "issuer"
+        <*> v .: "period"
+
+-- | Returns the key associated with the given key name
+getKey :: VaultConnection Authenticated -> VaultMountedPath -> Text -> IO Key
+getKey conn path = fmap unDataWrapper
+    . runVaultRequestAuthenticated conn
+    . newGetRequest
+    . mkPathWithKey KeysNamespace path
+
+-- | Represents a list of key names
+newtype KeyNames = KeyNames {unKeyNames :: [Text]}
+    deriving (Show, Eq)
+
+instance FromJSON KeyNames where
+    parseJSON = withObject "KeyNames" $ fmap KeyNames . (.: "keys")
+
+-- | Returns a list of TOTP keys stored in the given mount path
+listKeys :: VaultConnection Authenticated -> VaultMountedPath -> IO KeyNames
```

Comment:
It seems like we could produce `[Text]` instead of `KeyNames` here.

Reply:
a1264907164eaf944749486033a3e9c1df1c6bf5